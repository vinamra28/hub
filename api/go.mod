module github.com/tektoncd/hub/api

go 1.14

require (
	github.com/Azure/azure-sdk-for-go v50.2.0+incompatible // indirect
	github.com/Azure/go-autorest/autorest v0.11.17 // indirect
	github.com/Azure/go-autorest/autorest/adal v0.9.10 // indirect
	github.com/Azure/go-autorest/autorest/to v0.4.0 // indirect
	github.com/Azure/go-autorest/autorest/validation v0.3.1 // indirect
	github.com/aws/aws-sdk-go v1.37.1 // indirect
	github.com/dgrijalva/jwt-go v3.2.0+incompatible
	github.com/docker/cli v20.10.2+incompatible // indirect
	github.com/docker/docker v20.10.2+incompatible // indirect
	github.com/emicklei/go-restful v2.15.0+incompatible // indirect
	github.com/go-gormigrate/gormigrate/v2 v2.0.0
	github.com/go-openapi/spec v0.20.2 // indirect
	github.com/go-testfixtures/testfixtures/v3 v3.2.0
	github.com/gogo/protobuf v1.3.2 // indirect
	github.com/google/go-containerregistry/pkg/authn/k8schain v0.0.0-20210129212729-5c4818de4025 // indirect
	github.com/google/go-github v17.0.0+incompatible
	github.com/google/gofuzz v1.2.0 // indirect
	github.com/googleapis/gnostic v0.5.3 // indirect
	github.com/hashicorp/go-version v1.2.0
	github.com/ikawaha/goahttpcheck v1.3.1
	github.com/joho/godotenv v1.3.0
	github.com/mitchellh/go-homedir v1.1.0
	github.com/sirupsen/logrus v1.7.0 // indirect
	github.com/spf13/cobra v1.1.1
	github.com/spf13/viper v1.7.0
	github.com/stretchr/testify v1.6.1
	github.com/tektoncd/cli v0.16.0
	github.com/tektoncd/pipeline v0.20.1
	go.uber.org/zap v1.16.0
	goa.design/goa/v3 v3.2.2
	goa.design/plugins/v3 v3.1.3
	golang.org/x/crypto v0.0.0-20201221181555-eec23a3978ad
	golang.org/x/mod v0.4.1 // indirect
	golang.org/x/oauth2 v0.0.0-20210126194326-f9ce19ea3013
	golang.org/x/term v0.0.0-20201210144234-2321bbc49cbf // indirect
	golang.org/x/time v0.0.0-20201208040808-7e3f01d25324 // indirect
	golang.org/x/tools v0.1.0 // indirect
	gopkg.in/h2non/gock.v1 v1.0.15
	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b // indirect
	gorm.io/driver/postgres v1.0.2
	gorm.io/gorm v1.20.7
	gotest.tools/v3 v3.0.2
	k8s.io/api v0.19.7
	k8s.io/apimachinery v0.19.7
	k8s.io/client-go v11.0.1-0.20190805182717-6502b5e7b1b5+incompatible
	k8s.io/utils v0.0.0-20210111153108-fddb29f9d009 // indirect
	knative.dev/pkg v0.0.0-20210203171706-6045ed499615
	maze.io/x/duration v0.0.0-20160924141736-faac084b6075
)

replace k8s.io/client-go => k8s.io/client-go v0.19.7
