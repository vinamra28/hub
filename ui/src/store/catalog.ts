import { types, Instance } from 'mobx-state-tree';

export const CatalogType = types
  .model({
    id: types.integer,
    name: types.string,
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type ICatalogType = Instance<typeof CatalogType>;
export type ICatalogStore = Instance<typeof CatalogStore>;

export const CatalogStore = types
  .model('CatalogStore', {
    list: types.array(CatalogType)
  })

  .views((self) => ({
    get count(): number {
      return self.list.length;
    },
    get catalogs(): Array<string> {
      const { list } = self;
      return list
        .filter((type: ICatalogType) => type.selected)
        .map((type: ICatalogType) => type.name);
    }
  }))

  .actions((self) => ({
    add(item: ICatalogType) {
      self.list.push(item);
    },
    clear() {
      self.list.map((item: ICatalogType) => (item.selected = false));
    }
  }));
