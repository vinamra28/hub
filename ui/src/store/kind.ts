import { types, Instance } from 'mobx-state-tree';

export const Kind = types
  .model('Kind', {
    id: types.integer,
    name: types.string,
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type IKind = Instance<typeof Kind>;
export type IKindStore = Instance<typeof KindStore>;

export const KindStore = types
  .model('KindStore', {
    list: types.array(Kind)
  })

  .views((self) => ({
    get count(): number {
      return self.list.length;
    },
    get kinds(): Array<string> {
      const { list } = self;
      return list.filter((type: IKind) => type.selected).map((type: IKind) => type.name);
    }
  }))

  .actions((self) => ({
    add(item: IKind) {
      self.list.push(item);
    },
    clear() {
      self.list.map((item: IKind) => (item.selected = false));
    }
  }));
