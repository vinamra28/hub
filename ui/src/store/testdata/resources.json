{
  "data": [
    {
      "id": 5,
      "name": "buildah",
      "catalog": {
        "id": 1,
        "type": "official"
      },
      "kind": "Task",
      "latestVersion": {
        "id": 5,
        "version": "0.1",
        "displayName": "buildah",
        "description": "Buildah task builds source into a container image and then pushes it to a container registry.\nBuildah Task builds source into a container image using Project Atomic's Buildah build tool.It uses Buildah's support for building from Dockerfiles, using its buildah bud command.This command executes the directives in the Dockerfile to assemble a container image, then pushes that image to a container registry.",
        "minPipelinesVersion": "",
        "rawURL": "https://raw.githubusercontent.com/Pipelines-Marketplace/catalog/master/task/buildah/0.1/buildah.yaml",
        "webURL": "https://github.com/Pipelines-Marketplace/catalog/tree/master/task/buildah/0.1/buildah.yaml",
        "updatedAt": "2020-07-17 12:26:26.835302 +0000 UTC"
      },
      "tags": [
        {
          "id": 1,
          "name": "image-build"
        }
      ],
      "rating": 5
    },
    {
      "id": 4,
      "name": "build-push-gke-deploy",
      "catalog": {
        "id": 1,
        "type": "official"
      },
      "kind": "Task",
      "latestVersion": {
        "id": 4,
        "version": "0.1",
        "displayName": "",
        "description": "This Pipeline builds, pushes, and deploys your application to a Google Kubernetes Engine cluster using gke-deploy.",
        "minPipelinesVersion": "",
        "rawURL": "https://raw.githubusercontent.com/Pipelines-Marketplace/catalog/master/task/build-push-gke-deploy/0.1/build-push-gke-deploy.yaml",
        "webURL": "https://github.com/Pipelines-Marketplace/catalog/tree/master/task/build-push-gke-deploy/0.1/build-push-gke-deploy.yaml",
        "updatedAt": "2020-07-17 12:26:26.831271 +0000 UTC"
      },
      "tags": [
        {
          "id": 10,
          "name": "deploy"
        }
      ],
      "rating": 5
    },
    {
      "id": 6,
      "name": "buildkit",
      "catalog": {
        "id": 1,
        "type": "official"
      },
      "kind": "Task",
      "latestVersion": {
        "id": 6,
        "version": "0.1",
        "displayName": "buildkit",
        "description": "This Task builds source into a container image using Moby BuildKit.",
        "minPipelinesVersion": "",
        "rawURL": "https://raw.githubusercontent.com/Pipelines-Marketplace/catalog/master/task/buildkit/0.1/buildkit.yaml",
        "webURL": "https://github.com/Pipelines-Marketplace/catalog/tree/master/task/buildkit/0.1/buildkit.yaml",
        "updatedAt": "2020-07-17 12:26:26.839271 +0000 UTC"
      },
      "tags": [
        {
          "id": 1,
          "name": "image-build"
        }
      ],
      "rating": 4
    },
    {
      "id": 7,
      "name": "buildkit-daemonless",
      "catalog": {
        "id": 1,
        "type": "official"
      },
      "kind": "Task",
      "latestVersion": {
        "id": 7,
        "version": "0.1",
        "displayName": "buildkit daemonless",
        "description": "This Task builds source into a container image using Moby BuildKit.\nThis buildkit-daemonless Task is similar to buildkit but does not need creating Secret, Deployment, and Service resources for setting up the buildkitd daemon cluster.",
        "minPipelinesVersion": "",
        "rawURL": "https://raw.githubusercontent.com/Pipelines-Marketplace/catalog/master/task/buildkit-daemonless/0.1/buildkit-daemonless.yaml",
        "webURL": "https://github.com/Pipelines-Marketplace/catalog/tree/master/task/buildkit-daemonless/0.1/buildkit-daemonless.yaml",
        "updatedAt": "2020-07-17 12:26:26.843821 +0000 UTC"
      },
      "tags": [
        {
          "id": 1,
          "name": "image-build"
        }
      ],
      "rating": 4
    },
    {
      "id": 2,
      "name": "argocd",
      "catalog": {
        "id": 1,
        "type": "official"
      },
      "kind": "Task",
      "latestVersion": {
        "id": 2,
        "version": "0.1",
        "displayName": "argocd",
        "description": "This task syncs (deploys) an Argo CD application and waits for it to be healthy.\nTo do so, it requires the address of the Argo CD server and some form of authentication either a username/password or an authentication token.",
        "minPipelinesVersion": "",
        "rawURL": "https://raw.githubusercontent.com/Pipelines-Marketplace/catalog/master/task/argocd/0.1/argocd.yaml",
        "webURL": "https://github.com/Pipelines-Marketplace/catalog/tree/master/task/argocd/0.1/argocd.yaml",
        "updatedAt": "2020-07-17 12:26:26.822315 +0000 UTC"
      },
      "tags": [
        {
          "id": 10,
          "name": "deploy"
        }
      ],
      "rating": 3.5
    }
  ]
}
