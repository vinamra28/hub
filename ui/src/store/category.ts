import { types, getEnv, flow, Instance } from 'mobx-state-tree';
import { Api } from '../api';

export const Tag = types.model('Tags', {
  id: types.integer,
  name: types.string
});

export const Category = types
  .model('Category', {
    id: types.number,
    name: types.string,
    tags: types.array(Tag),
    selected: false
  })
  .actions((self) => ({
    toggle() {
      self.selected = !self.selected;
    }
  }));

export type ICategory = Instance<typeof Category>;
export type ITag = Instance<typeof Tag>;

export const CategoryStore = types
  .model('CategoryStore', {
    items: types.array(Category),
    isLoading: true,
    err: ''
  })

  .views((self) => ({
    get api(): Api {
      return getEnv(self).api;
    },

    get count() {
      return self.items.length;
    },

    get tags() {
      return self.items
        .filter((c) => c.selected)
        .reduce((acc: string[], c: ICategory) => [...acc, ...c.tags.map((t) => t.name)], []);
    }
  }))

  .actions((self) => ({
    add(item: ICategory) {
      self.items.push(item);
    },

    setLoading(l: boolean) {
      self.isLoading = l;
    },

    clear() {
      self.items.map((c: ICategory) => (c.selected = false));
    }
  }))

  .actions((self) => ({
    load: flow(function* () {
      try {
        self.setLoading(true);
        const { api } = self;
        const json = yield api.categories();
        json.data.forEach((item: ICategory) => self.add(item));
      } catch (err) {
        self.err = err.toString();
      }
      self.setLoading(false);
    })
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    }
  }));

export type ICategoryStore = Instance<typeof CategoryStore>;
