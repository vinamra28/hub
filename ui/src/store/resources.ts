import { types, getEnv, flow, Instance } from 'mobx-state-tree';
import fuzzysort from 'fuzzysort';
import { Api } from '../api';
import { ICategoryStore, Tag, ITag } from './category';
import { IKindStore, Kind } from './kind';
import { ICatalogStore, CatalogType } from './catalog';

const Catalog = types.model('Catalog', {
  id: types.number,
  type: types.string
});

const ResourceVersion = types.model('Version', {
  id: types.number,
  version: types.optional(types.string, '0.1'),
  displayName: types.string,
  description: types.string,
  minPipelinesVersion: types.optional(types.string, ''),
  rawURL: types.string,
  webURL: types.string,
  updatedAt: types.string
});

export const Resource = types.model('Resource', {
  id: types.number,
  kind: types.string,
  name: types.string,
  catalog: Catalog,
  latestVersion: ResourceVersion,
  tags: types.optional(types.array(Tag), []),
  rating: 0
});

export type IResource = Instance<typeof Resource>;

export const ResourceStore = types
  .model('ResourceStore', {
    resources: types.array(Resource),
    isLoading: true,
    searchInput: types.optional(types.string, ''),
    err: ''
  })

  .views((self) => ({
    get api(): Api {
      return getEnv(self).api;
    },
    get kindStore(): IKindStore {
      return getEnv(self).kindStore;
    },
    get catalogStore(): ICatalogStore {
      return getEnv(self).catalogStore;
    },
    get categoryStore(): ICategoryStore {
      return getEnv(self).categoryStore;
    },
    get count(): number {
      return self.resources.length;
    },
    get resourceList(): Array<IResource> {
      const { resources } = self;
      return resources;
    }
  }))

  .actions((self) => ({
    add(item: IResource) {
      self.resources.push(item);
    },
    setLoading(l: boolean) {
      self.isLoading = l;
    },
    setSearchInput(text: string) {
      self.searchInput = text;
    }
  }))

  .actions((self) => ({
    load: flow(function* () {
      try {
        self.setLoading(true);

        const catalogData = new Set();
        const kindData = new Set();

        const { api } = self;
        const json = yield api.resources();

        json.data.forEach((item: IResource) => {
          self.add(item);
          catalogData.add(item.catalog.type);
          kindData.add(item.kind);
        });

        Array.from(kindData).forEach((kindName, index) => {
          self.kindStore.add(Kind.create({ id: index, name: kindName as string }));
        });

        Array.from(catalogData).forEach((catalogName, index) => {
          self.catalogStore.add(CatalogType.create({ id: index, name: catalogName as string }));
        });
      } catch (err) {
        self.err = err.toString();
      }
      self.setLoading(false);
    }),
    clearAll() {
      self.kindStore.clear();
      self.catalogStore.clear();
      self.categoryStore.clear();
      self.setSearchInput('');
    }
  }))

  .views((self) => ({
    get list(): Array<IResource> {
      const { resources } = self;
      const { searchInput } = self;

      const kind = new Set(self.kindStore.kinds);
      const tag = new Set(self.categoryStore.tags);
      const catalog = new Set(self.catalogStore.catalogs);

      const filterResources = resources.filter(
        (r: IResource) =>
          (tag.size > 0 ? r.tags.some((t: ITag) => tag.has(t.name)) : true) &&
          (catalog.size > 0 ? catalog.has(r.catalog.type) : true) &&
          (kind.size > 0 ? kind.has(r.kind) : true)
      );

      return searchInput !== ''
        ? fuzzysort
            .go(searchInput, filterResources, {
              keys: ['name', 'displayName']
            })
            .map((resource: Fuzzysort.KeysResult<IResource>) => resource.obj)
        : filterResources;
    }
  }))

  .actions((self) => ({
    afterCreate() {
      self.load();
    }
  }));
