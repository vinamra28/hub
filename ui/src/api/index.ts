import axios from 'axios';
import { API_URL } from '../config/constants';
import { ICategory } from '../store/category';
import { IResource } from '../store/resources';

export interface Api {
  categories(): Promise<ICategory>;
  resources(): Promise<IResource>;
}

export class Hub implements Api {
  async resources() {
    try {
      return axios.get(`${API_URL}/resources`);
    } catch (err) {
      return err.response;
    }
  }

  async categories() {
    try {
      return axios.get(`${API_URL}/categories`);
    } catch (err) {
      return err.response;
    }
  }
}
