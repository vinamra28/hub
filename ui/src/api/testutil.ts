import { Api } from './';
import * as fs from 'fs';
import { ICategory } from '../store/category';
import { IResource } from '../store/resources';

export class FakeHub implements Api {
  dataDir: string;

  constructor(dataDir: string) {
    this.dataDir = dataDir;
  }

  async resources() {
    const data = `${this.dataDir}/resources.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<IResource>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }

  async categories() {
    const data = `${this.dataDir}/categories.json`;

    const ret = () => JSON.parse(fs.readFileSync(data).toString());
    return new Promise<ICategory>((resolve) => {
      setTimeout(() => resolve(ret()), 1000);
    });
  }
}
