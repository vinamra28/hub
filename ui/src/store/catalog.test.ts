import { getSnapshot } from 'mobx-state-tree';
import { when } from 'mobx';
import { CatalogStore, CatalogType } from './catalog';
import { CategoryStore } from './category';
import { FakeHub } from '../api/testutil';
import { KindStore } from './kind';
import { ResourceStore } from './resources';

const TESTDATA_DIR = `${__dirname}/testdata`;
const api = new FakeHub(TESTDATA_DIR);

describe('Catalog', () => {
  it('can create a catalog object', (done) => {
    const store = CatalogType.create({
      id: 3,
      name: 'official'
    });
    expect(store.name).toBe('official');

    done();
  });
});

describe('Catalog Store', () => {
  it('can create a store', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        kindStore: KindStore.create({}),
        catalogStore: CatalogStore.create({}),
        categoryStore: CategoryStore.create({}, { api })
      }
    );

    when(
      () => !store.isLoading,
      () => {
        expect(store.isLoading).toBe(false);

        expect(store.catalogStore.count).toBe(1);
        expect(store.catalogStore.list[0].name).toBe('official');
        expect(getSnapshot(store.catalogStore.list)).toMatchSnapshot();

        done();
      }
    );
  });

  it('can toggle the selected catalog', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        kindStore: KindStore.create({}),
        catalogStore: CatalogStore.create({}),
        categoryStore: CategoryStore.create({}, { api })
      }
    );

    when(
      () => !store.isLoading,
      () => {
        expect(store.count).toBe(5);
        expect(store.isLoading).toBe(false);

        store.catalogStore.list[0].toggle();
        expect(store.catalogStore.list[0].selected).toBe(true);

        done();
      }
    );
  });

  it('can return the catalogType which are selected', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        kindStore: KindStore.create({}),
        catalogStore: CatalogStore.create({}),
        categoryStore: CategoryStore.create({}, { api })
      }
    );

    when(
      () => !store.isLoading,
      () => {
        expect(store.isLoading).toBe(false);

        store.catalogStore.list[0].toggle();

        expect(store.catalogStore.catalogs[0]).toBe('official');

        done();
      }
    );
  });

  it('can clear all selected catalog', (done) => {
    const store = ResourceStore.create(
      {},
      {
        api,
        kindStore: KindStore.create({}),
        catalogStore: CatalogStore.create({}),
        categoryStore: CategoryStore.create({}, { api })
      }
    );

    when(
      () => !store.isLoading,
      () => {
        expect(store.isLoading).toBe(false);

        store.catalogStore.list[0].toggle();
        store.catalogStore.clear();

        expect(store.catalogStore.list).toEqual(
          expect.arrayContaining([
            expect.objectContaining({
              selected: false
            })
          ])
        );

        done();
      }
    );
  });
});
