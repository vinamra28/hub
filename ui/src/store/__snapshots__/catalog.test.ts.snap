// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog Store can create a store 1`] = `
Array [
  Object {
    "id": 0,
    "name": "official",
    "selected": false,
  },
]
`;
